notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+      ]
+     },
+     "execution_count": 18,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql='''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, \n",
+    "  u.v as age_group, \n",
+    "  count(distinct c.client_ip) as distint_ips\n",
+    "FROM clicks c \n",
+    "     INNER JOIN lookup.user_age_group u\n",
+    "        ON c.user_id=u.k\n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, 2\n",
+    "ORDER BY 3 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"simple_join_lookup\",\"results\":result})\n",
+    "print_timestamp()\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dc877650-7728-44c7-9aa9-5f67ac606503",
+   "metadata": {},
+   "source": [
+    "... and the second query using the `lookup`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "id": "0aee9a0e-fa53-4166-a756-8b24c13ccb2d",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:550.8509 ms   min:519.703 ms  max:608.721 ms\n",
+      "2023-10-26 01:01:49.870030 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 1, 49, 870030)"
+      ]
+     },
+     "execution_count": 19,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql = '''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, c.user_id,\n",
+    "  u.v as age_group, \n",
+    "  count(1) as user_clicks,\n",
+    "  count(1)*100.0/agg.clicks as age_group_contrib_pct\n",
+    "FROM clicks c\n",
+    "     INNER JOIN lookup.user_age_group u ON c.user_id = u.k\n",
+    "     INNER JOIN\n",
+    "     (\n",
+    "         SELECT TIME_FLOOR(cc.__time, 'P1D') as _date,\n",
+    "                uu.v as age_group,\n",
+    "                COUNT(1) as clicks\n",
+    "           FROM clicks cc\n",
+    "           INNER JOIN lookup.user_age_group uu ON cc.user_id=uu.k\n",
+    "           GROUP BY 1, 2\n",
+    "     ) agg\n",
+    "        ON agg._date = TIME_FLOOR(c.__time, 'P1D') AND agg.age_group=u.v \n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, 2, 3, agg.clicks \n",
+    "ORDER BY 1, 5 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"subquery_join_lookup\",\"results\":result})\n",
+    "print_timestamp()\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "62205ce8-a1c4-4256-81b3-f6c287e4017a",
+   "metadata": {},
+   "source": [
+    "### Query Pre-Joined Data\n",
+    "Avoiding the join altogether is usually the best option.\n",
```

Comment:
This is advisory.

Reply:
It's true. All I'm trying to say here, is that if you can avoid the join, avoid it. This is the guidance that I've seen Gian give many times. It makes sense for performance and for overall CPU/Memory usage at query time.